### PR TITLE
sagittarius-scheme: init at 0.9.6

### PR DESCRIPTION
**Pull request by wahjava**
###### Motivation for this change

Add package for Sagittarius Scheme

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -85,6 +85,15 @@
     github = "baldo";
     name = "Andreas Baldeau";
   };
+  abbe = {
```
Comment (aanderse) [inline maintainers/maintainer-list.nix]: Please add yourself to `maintainer-list.nix` in a separate commit first.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
+{ stdenv
+, fetchurl
+, cmake
+, pkgconfig
+, libffi
+, boehmgc
+, openssl
+, zlib
+, odbc ? true
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: It is more common to write this as `odbcSupport`.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  src = fetchurl {
+    url = "https://bitbucket.org/ktakashi/sagittarius-scheme/downloads/sagittarius-${version}.tar.gz";
+    sha256 = "03nvvvfd4gdlvq244zpnikxxajp6w8jj3ymw4bcq83x7zilb2imr";
+  };
+  preBuild = ''
+           export LD_LIBRARY_PATH="$(pwd)/build"
+  '';
+  nativeBuildInputs = [ pkgconfig cmake ];
+
+  buildInputs = builtins.concatLists [[ libffi boehmgc openssl zlib ]  odbcDeps];
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: This should be rewritten as
```
buildInputs = [ libffi boehmgc openssl zlib ] + lib.optional odbcSupport libiodbc;
```

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  meta = with stdenv.lib; {
+    description = "Sagittarius Scheme is an R6RS/R7RS Scheme system";
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: From the nixpkgs manual:
> Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+     -  Common Lisp like reader macro.
+     -  Cryptographic libraries.
+     -  Customisable cipher and hash algorithm.
+     -  Custom codec mechanism.
+     -  CL like keyword lambda syntax (taken from Gauche).
+     -  Constant definition form. (define-constant form).
+     -  Builtin regular expression
+     -  mostly works O(n)
+     -   Replaceable reader
+    '';
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: Please be consistent with your identation. Specifically the first line of `longDescription` should be indented two additional spaces.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    homepage = https://bitbucket.org/ktakashi/sagittarius-scheme;
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: Please quote as quoting urls will eventually become mandatory.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    license = licenses.bsd2;
+    platforms = platforms.x86;
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: After briefly reading the `homepage` it seems `platforms.all` would be more appropriate.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+, openssl
+, zlib
+, odbc ? true
+, libiodbc
+}:
+
+let version = "0.9.6";
+    odbcDeps = if odbc then [ libiodbc ] else [];
+in stdenv.mkDerivation {
+  name = "sagittarius-scheme-${version}";
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: Please replace with `pname = "sagittarius-scheme";` as `name = "${pname}-${version}";` is already implied.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+, libffi
+, boehmgc
+, openssl
+, zlib
+, odbc ? true
+, libiodbc
+}:
+
+let version = "0.9.6";
+    odbcDeps = if odbc then [ libiodbc ] else [];
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: This line is not required. See `buildInputs` comment below.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+, pkgconfig
+, libffi
+, boehmgc
+, openssl
+, zlib
+, odbc ? true
+, libiodbc
+}:
+
+let version = "0.9.6";
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: Please move `version` directly into the derivation.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    odbcDeps = if odbc then [ libiodbc ] else [];
+in stdenv.mkDerivation {
+  name = "sagittarius-scheme-${version}";
+  src = fetchurl {
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: `fetchurl` -> `fetchFromBitbucket`. See https://nixos.org/nixpkgs/manual/ for details.

File: pkgs/top-level/all-packages.nix [hunk 12]
```diff
@@ -8202,6 +8202,8 @@ in
     inherit (darwin.apple_sdk.frameworks) CoreServices Security;
   };
 
+  sagittariusScheme = callPackage ../development/compilers/sagittarius-scheme {};
```
Comment (aanderse) [inline pkgs/top-level/all-packages.nix]: `sagittariusScheme` -> `sagittarius-scheme`.

File: pkgs/development/compilers/sagittarius-scheme/default.nix [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    url = "https://bitbucket.org/ktakashi/sagittarius-scheme/downloads/sagittarius-${version}.tar.gz";
+    sha256 = "03nvvvfd4gdlvq244zpnikxxajp6w8jj3ymw4bcq83x7zilb2imr";
+  };
+  preBuild = ''
+           export LD_LIBRARY_PATH="$(pwd)/build"
```
Comment (aanderse) [inline pkgs/development/compilers/sagittarius-scheme/default.nix]: Why is this required?

**Review by aanderse — CHANGES_REQUESTED** [review]
@wahjava Thanks for contributing to nixpkgs and going so far as to adding yourself to the `maintainers-list.nix` file! We all appreciate your contribution :tada: 

I have left some comments I hope you find useful. If you have any questions at all please do not hesitate to ask.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (wahjava) [reply maintainers/maintainer-list.nix]: Sure, I can split in two commits. Do I need to do open another pull request for `maintainer-list` change ?

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 13 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Sorry, I will add a comment in the file, though the reason is we (Nix's cmake tooling) skip adding rpaths during build (`-DCMAKE_SKIP_BUILD_RPATH=ON`) and build of Sagittarius involves building a binary executable `sagittarius` (which is linked to its own `libsagittarius.so`) and is then required to be executed to process couple of Scheme files, and dies due to not being able to resolve `libsagittarius.so` library (in the absence of `LD_LIBRARY_PATH` or `rpath`).

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 11 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: The distribution tarball available includes the auto-generated .c files which are not included in the repository. Building directly from repository require a Sagittarius bootstrap which is not needed if you build from a distribution tarball.

Let me know if this is a good enough excuse to keep it this way.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 10 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 9 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 4 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 3 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 2 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/top-level/all-packages.nix (hunk 12 above)
Comment (wahjava) [reply pkgs/top-level/all-packages.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 5 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 7 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 6 above)
Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Done.

**Comment by wahjava** [comment]
Thanks for your advice. I have implemented them except for `fetchBitBucket`  change as mentioned above. While trying to avoid forced commits, I ended up deleting the branch, and re-pushing new commits to a branch with same name, and to my surprise it turns out that resulted in closing this pull request.

Following are the new commits created by me on the branch:

e55179557abf51754f78f91fbaf7cc12099807b7
ec3e2a94038e21f11672d8f2bd7bd8c8217491d8

I wonder if it's possible for you to re-open this pull-request, or you like me to submit a new PR referencing this one.

Sorry for generating more work for you.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (aanderse) [reply maintainers/maintainer-list.nix]: The same PR is fine, just a separate commit please.

**Comment by aanderse** [comment]
Thanks for your work on this. I also seem to be unable to reopen the PR. I wonder if this is the resolution: https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c?

**Comment by wahjava** [comment]
Re-opening it :)

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 11 above)
Comment (aanderse) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: This is an entirely valid reason. Thanks for clarifying.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 5 above)
Comment (aanderse) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Sorry, I guess what I meant to say is that all lines (including the first) should be indented an additional 2 spaces, like so:
```
    longDescription = ''
      Sagittarius Scheme is a free Scheme implementation supporting
      R6RS/R7RS specification.
      Features:
      -  Builtin CLOS.
      -  Common Lisp like reader macro.
      -  Cryptographic libraries.
      -  Customisable cipher and hash algorithm.
      -  Custom codec mechanism.
      -  CL like keyword lambda syntax (taken from Gauche).
      -  Constant definition form. (define-constant form).
      -  Builtin regular expression
      -  mostly works O(n)
      -  Replaceable reader
    '';
```

Also note how the last entry `Replaceable reader` has an additional space which should not be there.

Comment (wahjava) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: Sorry, I got confused by Debian's convention of indenting first line. Anyways this should now be taken care of by the most recent commit.

File: pkgs/development/compilers/sagittarius-scheme/default.nix (hunk 13 above)
Comment (aanderse) [reply pkgs/development/compilers/sagittarius-scheme/default.nix]: OK. Please keep indentation consistent at 2 spaces per level.

**Review by aanderse — APPROVED** [review]
Overall looks good to me. I'll wait for someone with domain specific knowledge to either approve or merge, though. If you don't hear anything on this PR in the next few days ping me and we can actively look for someone.

Thanks again @wahjava!

**Comment by aanderse** [comment]
@GrahamcOfBorg build sagittarius-scheme

**Comment by aanderse** [comment]
@wahjava If you're able to fix the `darwin` build issue that is great, but if not please update `platforms` accordingly.

**Comment by wahjava** [comment]
> @wahjava If you're able to fix the `darwin` build issue that is great, but if not please update `platforms` accordingly.

I can try. I think I know what needs to be fixed, need `LD_LIBRARY_PATH` equivalent for for Darwin platform. I will ask in channel, if someone can try it on their Darwin builders.

**Comment by wahjava** [comment]
@aanderse pushed new changes, in 707c5a4, which I believe should fix the build for `darwin` platform.

**Comment by wahjava** [comment]
> @aanderse pushed new changes in [707c5a4](https://github.com/NixOS/nixpkgs/commit/707c5a4faae6e45e39ad60997daa20b9ed1ef10d), which I believe should fix the build for `darwin` platform.

sorry seems like like a wrong commit on my end 8c32ecd should really fix.

**Comment by aanderse** [comment]
Oh no! The `preBuild` indentation came back!

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (7c6f434c) [inline maintainers/maintainer-list.nix]: Just curious — the maintainer nickname doesn't coincide with anything in the maintainer entry. Should we add some fields to the schema? (Is it your IRC nickname?)

Comment (wahjava) [reply maintainers/maintainer-list.nix]: Yes, it's.

**Comment by 7c6f434c** [comment]
@aanderse I have built enough Scheme implementations to judge that domain knowledge is _not_ actually needed for evaluating this expression!

**Pull request merged**